src/engine/Join.cpp

```diff
@@ -409,6 +417,65 @@ void Join::join(const IdTable& a, ColumnIndex jc1, const IdTable& b,
              << ", size = " << result->size() << "\n";
 }
 
+// _____________________________________________________________________________
+Result::Generator Join::yieldOnCallbackCalled(
+    ad_utility::InvocableWithExactReturnType<
+        Result::IdTableVocabPair,
+        std::function<void(IdTable&, LocalVocab&)>> auto action,
+    std::vector<ColumnIndex> permutation) const {
+  std::atomic_flag write = true;
+  std::variant<std::monostate, Result::IdTableVocabPair, std::exception_ptr>
+      storage;
+  ad_utility::JThread thread{[&write, &storage, &action, &permutation]() {
+    auto writeValue = [&write, &storage](auto value) noexcept {
+      storage = std::move(value);
+      write.clear();
+      write.notify_one();
+    };
+    auto writeValueAndWait = [&permutation, &write,
+                              &writeValue](Result::IdTableVocabPair value) {
+      AD_CORRECTNESS_CHECK(write.test());
+      if (!permutation.empty()) {
+        value.idTable_.setColumnSubset(permutation);
+      }
```

Comment:
Make the free helper function `applyPermutation(permutation, value.idTable_)` to save two lines in this complex lambda + reduce some code duplication (it reappears below).